Wallet Developer's guide (#307)
ethereum-optimism/community-hub: src/docs/developers/by-role/wallet-dev.md

---
title: Supporting Optimism in Your Wallet
lang: en-US
---

# {{ $frontmatter.title }}

## Handling multiple chains

There are two ways for a wallet to handle being connected to multiple chains:

> 1. Connect to one chain at a time, the way MetaMask does it. Let the user switch chains either from your user interface, or through a user interface.

Reviewer: Fix this sentence -> Let the user switch chains either from your user interface, or through a user interface.

1. Connect to all chains all the time, the way Rainbow does it. Display ETH and token balances on all chains, and ask which chain to use when connecting to a dapp.

### Switching chains

> A user can add or swithc into a chain using either the wallet's user interface or a dapp's user interface. To enable switching on a dapp, the wallet needs to handle these requests:

Reviewer: typo -> swithc

- [`wallet_addEthereumChain`](https://eips.ethereum.org/EIPS/eip-3085)

> - [`wallet_switchEthereumChain`](https://eips.ethereum.org/EIPS/eip-3326)

Reviewer: Open any external links in a new tab.

Author: That is handled by Vuepress. If you look at our other docs (for example, https://raw.githubusercontent.com/ethereum-optimism/community-hub/main/src/docs/infra/monitoring.md), all external links are in this format.

## Connecting to Optimism

> You connect to Optimism the same way you do to Ethereum, by connecting to a JSON RPC endpoint. However, the currently supported fork for Optimism is [Berlin](https://eth.wiki/roadmap/berlin), not [London](https://eth.wiki/roadmap/london), so EIP-1559 transactions are not supported yet.

Reviewer: Open external links in new tab.

### Endpoints

> [Click here for the Optimism endpoints](../../infra/networks.md). You can choose between our public endpoints, which are rate limited, and [endpoints from infrastructure providers](../../infra/networks.md#rpc-endpoints).

Reviewer 2: Just looked through these links - I think we should add a note in the infra page where it says:
> We also support and maintain the following public RPC endpoints. However, we highly encourage you to set up a private RPC endpoint with any of the above mentioned providers instead. These public endpoints should not be used in production systems.

We ought to include details around "These public endpoints are IP rate limited to ___ requests per IP address per minute".

Author: Apparently they are not rate limited right now. Should I say they are?

Reviewer 2: Believe we resolved this copy in slack

### Token addresses

> The ERC-20 contract address for a token on Optimism may be different from the address for the same token on Ethereum. [The list of tokens and their addresses is here](https://static.optimism.io/optimism.tokenlist.json).

Reviewer 2: can also link to: https://tokenlists.org/token-list?url=https://static.optimism.io/optimism.tokenlist.json

Author: Are you sure? That list isn't really set up for multi-chain tokens (see https://linear.app/optimism/issue/INT-440/the-token-list-ui-is-broken).

For example, looking at the **SNX** token, we get these addresses:

| ChainID | Network | Address |
| -: | - | - |
| 1  | Ethereum    | 0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f |
| 10 | Optimism    | 0x8700daec35af8ff88c16bdf0418774cb3d7599b4
| 42 | Kovan (test network) | 0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f
| 69 | Optimistic Kovan (test network) | 0x0064A673267696049938AA47595dD0B3C2e705A1

To get the total SNX balance of a user that uses Optimism you need to:

1. Connect to a standard Ethereum endpoint and send a `balanceOf` query to address `0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f`.
1. Connect to an Optimism endpoint and send a `balanceOf` query to address `0x8700daec35af8ff88c16bdf0418774cb3d7599b4`.

### Transaction status

A transaction in Optimism can be in one of three states:

- **Sequencer Confirmed**: The transaction has been accepted by the Optimism sequencer on L2
- **Confirmed On-Chain**: The transaction has been written to Ethereum (L1). In that case of an L2->L1 transaction, this starts the seven day period until it can be finalized

> - **Finalized On-Chain**: This status is only application for L2->L1 transactions, and means that the transaction was finalized on L1 after the verification period has passed.

Reviewer 2: ```suggestion
- **Finalized On-Chain**: This status is only application for L2->L1 transactions, and means that the transaction was finalized on L1 after the seven day verification period has passed.
```

## Differences from Ethereum

> You can see a summary of the differences [here](../l2/differences.md). These are the differences that are most relevant to wallet developers:

Reviewer 2: Optional suggestion:  

```suggestion
We aim to be [EVM equivalent](https://medium.com/ethereum-optimism/introducing-evm-equivalence-5c2021deb306), meaning we aim to minimize the differences between Optimism and Ethereum. You can see a summary of the few differences between Optimism and Ethereum [here](../l2/differences.md). These are the differences that are most relevant to wallet developers:
```

### Transaction fees

> Most of the cost of an Optimism transaction is not the gas consumed by the transaction itself (which is priced in most cases at 0.001 gwei), but the cost of writing the transaction in Ethereum. That cost is deducted automatically from the user's balance on Optimism.

Reviewer 2: ```suggestion
Most of the cost of an Optimism transaction is not the gas consumed by the transaction itself (which is priced at 0.001 gwei when the chain is not congested), but the cost of writing the transaction in Ethereum. That cost is deducted automatically from the user's balance on Optimism.
```
We need to draw more attention to this section because this is the single diff that matters for wallet developers beyond supporting any other EVM equivalent chain (e.g. Polygon). We need to call out explicitly that wallet developers *need to add custom code* to how they normally display fees, since what they usually display is now only the L2 portion of the total fee their user will pay.

| ChainID | Network | Address |
| -: | - | - |
| 1  | Ethereum    | 0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f |
| 10 | Optimism    | 0x8700daec35af8ff88c16bdf0418774cb3d7599b4
| 42 | Kovan (test network) | 0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f
| 69 | Optimistic Kovan (test network) | 0x0064A673267696049938AA47595dD0B3C2e705A1

To get the total SNX balance of a user that uses Optimism you need to:

1. Connect to a standard Ethereum endpoint and send a `balanceOf` query to address `0xc011a73ee8576fb46f5e1c5751ca3b9fe0af2a6f`.
1. Connect to an Optimism endpoint and send a `balanceOf` query to address `0x8700daec35af8ff88c16bdf0418774cb3d7599b4`.

### Transaction status

A transaction in Optimism can be in one of three states:

- **Sequencer Confirmed**: The transaction has been accepted by the Optimism sequencer on L2
- **Confirmed On-Chain**: The transaction has been written to Ethereum (L1). In that case of an L2->L1 transaction, this starts the seven day period until it can be finalized
- **Finalized On-Chain**: This status is only application for L2->L1 transactions, and means that the transaction was finalized on L1 after the seven day verification period has passed.

> 

Reviewer 2: My question as a wallet developer would be "How do I track this state of transactions sent through my wallet?" I think the answer right now is probably that we need to wait for the JS SDK for this... 

For now, we can add something like:

"For now, we recommend wallets consider transactions final after they are Sequencer confirmed as there are no reorgs on the chain and there is currently not standard tooling for tracking when transactions are confirmed on chain. We have a JS SDK coming soon that will contain standard tooling to track these three states of each transaction "

## Differences from Ethereum

We aim to be [EVM equivalent](https://medium.com/ethereum-optimism/introducing-evm-equivalence-5c2021deb306), meaning we aim to minimize the differences between Optimism and Ethereum. You can see a summary of the few differences between Optimism and Ethereum [here](../l2/differences.md). These are the differences that are most relevant to wallet developers:

### Transaction fees

> Most of the cost of an Optimism transaction is not the gas consumed by the transaction itself (which is priced at 0.001 gwei when the chain is not congested), but the cost of writing the transaction in Ethereum. That cost is deducted automatically from the user's balance on Optimism.

Reviewer 2: Would change language from cost -> transaction fee

E.g. "The majority of an Optimism transaction's fee is not from the gas..."

Reposting my same comment from before:

We need to draw more attention to this section because this is the single diff that matters for wallet developers beyond supporting any other EVM equivalent chain (e.g. Polygon). We need to call out explicitly that wallet developers need to add custom code to how they normally display fees, since what they usually display is now only the L2 portion of the total fee their user will pay.

^I think we can accomplish this by adding a highlight or bold to ensure that wallet developers know that they need to add *custom code* here.
